APV versioning changes

dkozinski opened the pull request.

dkozinski committed:
Added reading library version from file in CMakeLists.txt

Signed-off-by: Dawid Kozinski <d.kozinski@samsung.com>

dkozinski committed:
Refactor version functions

- Implement versioning functions with conditional compilation
- Added oapv_version to generate version string using OAPV_VERSION_MAJOR, OAPV_VERSION_MINOR, and OAPV_VERSION_PATCH.
- Added oapv_libversion function to return library version using OAPV_LIB_VERSION_MAJOR, OAPV_LIB_VERSION_MINOR, and OAPV_LIB_VERSION_PATCH.

Signed-off-by: Dawid Kozinski <d.kozinski@samsung.com>

dkozinski committed:
Updated library version to 1.0.0

- Changed the version in library_version.txt to v1.0.0
- Updated the version definition in oapv.c.

Signed-off-by: Dawid Kozinski <d.kozinski@samsung.com>

dkozinski committed:
Update CPack version settings to use LIB_VERSION and added missing inc/oapv_version.h.in file

Signed-off-by: Dawid Kozinski <d.kozinski@samsung.com>

jamrial commented on pkgconfig/oapv.pc.in:
```diff
@@ -7,7 +7,7 @@ includedir=@CMAKE_INSTALL_PREFIX@/@CMAKE_INSTALL_INCLUDEDIR@/@LIB_NAME_BASE@
 Description: Advanced Professional Video Codec
 
-Version: @PROJECT_VERSION_MAJOR@.@PROJECT_VERSION_MINOR@.@PROJECT_VERSION_PATCH@
+Version: @LIB_VERSION_MAJOR@.@LIB_VERSION_MINOR@.@LIB_VERSION_PATCH@
```
No, this one is fine staying as project version. Pkg-config is about a package, not a specific library within it.

jamrial commented on inc/oapv_version.h.in:
```diff
@@ -47,4 +48,9 @@
+// Project version
+#define OAPV_VERSION_MAJOR @VERSION_MAJOR@
+#define OAPV_VERSION_MINOR @VERSION_MINOR@
+#define OAPV_VERSION_PATCH @VERSION_PATCH@
```
Wasn't it decided that the header would only have the library version defines?

dkozinski replied:
> Wasn't it decided that the header would only have the library version defines?

At the moment, we still need this. The function oapv_version() returns the project version and requires the definitions OAPV_VERSION_MAJOR, OAPV_VERSION_MINOR, and OAPV_VERSION_PATCH. 
Currently, we still need two versions of the function. 
aopv_version() is used by the reference applications and returns the project version. 
The other is oapv_libversion(), which returns the library version and could be used for integration with other projects like FFmpeg. 

We will consider simplify it.

jamrial replied:
Can't you add those defines in a different, also generated but not installed header?

Also, oapv_version() and oapv_libversion() should ideally return a const char*, or a strdup() output that the user needs to free otherwise.

dkozinski committed:
Exposed oapv_libversion() finction

Signed-off-by: Dawid Kozinski <d.kozinski@samsung.com>

jamrial commented on src/oapv.c:
```diff
@@ -2153,5 +2153,26 @@ int oapvd_info(void *au, int au_size, oapv_au_info_t *aui)
+static char *oapv_libver = "1.0.0";
+
 char * oapv_version() { return oapv_ver; }
+char * oapv_libversion() { return oapv_libver; }
```
I think this would be better as

```
unsigned oapv_libversion(void) {
    return (OAPV_LIB_VERSION_MAJOR << 16) |
           (OAPV_LIB_VERSION_MINOR <<  8) |
            OAPV_LIB_VERSION_PATCH;
}
```

And then adding the following to oapv_version.h

```
#define OAPV_LIB_MAJOR(v) (((v) >> 16) & 0xFF)
#define OAPV_LIB_MINOR(v) (((v) >>  8) & 0xFF)
#define OAPV_LIB_PATCH(v) ( (v)        & 0xFF)
```

This is a lot more library user friendly than a string for runtime version checks.
oapv_version() can remain a string just fine, but it should be const char*.

dkozinski committed:
Refactor version handling to use unsigned integers

- Updated the oapv_version and oapv_libversion functions to return unsigned integers instead of strings

Signed-off-by: Dawid Kozinski <d.kozinski@samsung.com>

jamrial commented on inc/oapv_config.h:
```diff
@@ -32,7 +32,7 @@
-// #define OAPV_LIBVERSION_HEADER // defined if the file oapv_libversion.h exists
-// #define OAPV_EXPORT_HEADER     // defined if the file oapv_exports.h exists
+#define OAPV_VERSION_HEADER // defined if the file oapv_libversion.h exists
+#define OAPV_EXPORT_HEADER  // defined if the file oapv_exports.h exists
```
What is the point of these defines? Why would those headers not exist if they are unconditionally installed?

Without this you'd be able to remove the ifdeffery in src/oapv.c and allow oapv_version() to safely return a const char*. And the reason a const char* is better for it than an int is because ideally you'd fetch the project version using git describe --tags first, and if not possible (Like in a tarball snapshot), then version.txt would be parsed.

kpchoi commented on inc/oapv.h:
```diff
@@ -662,9 +680,15 @@ OAPV_EXPORT int oapvd_decode(oapvd_t did, oapv_bitb_t *bitb, oapv_frms_t *ofrms,
+ * openapv project version
  *****************************************************************************/
-OAPV_EXPORT char *oapv_version();
+OAPV_EXPORT unsigned oapv_version();
```
Suggest to use ```unsigned int``` instead of just ```unsigned``` here

jamrial commented on src/oapv.c:
```diff
@@ -2153,5 +2153,26 @@ int oapvd_info(void *au, int au_size, oapv_au_info_t *aui)
+#else
+
 static char *oapv_ver = "0.1.13.1";
+static char *oapv_libver = "0.1.13.1";
```
It was decided in https://github.com/AcademySoftwareFoundation/openapv/pull/91#issuecomment-2875414654, and is already implemented. This PR at first reverted it by accident, but it was then rectified.

tuzm24 commented on src/oapv.c:
```diff
@@ -2153,5 +2153,24 @@ int oapvd_info(void *au, int au_size, oapv_au_info_t *aui)
+
+#else
+
+unsigned oapv_version() { 
+    return OAPV_GET_VERSION(0, 1, 13);
+}
+
+unsigned oapv_libversion() { 
+    return OAPV_GET_VERSION(1, 0, 0);
+}
+
+#endif
```
@dkozinski @jamrial 
Version numbering is the important. DO NOT try to decide it without maintainer's decision.
As I emphasized, it should be aligned to the market requirement.
And, I don't still fully understand any special reason why project and library version should be different in this project.

dkozinski committed:
Update versioning for project and library

- Refactor versioning strategy based on feedback from maintainer

Signed-off-by: Dawid Kozinski <d.kozinski@samsung.com>

dkozinski committed:
Removed obsolete library version checks from CMakeLists.txt

Signed-off-by: Dawid Kozinski <d.kozinski@samsung.com>

dkozinski committed:
Added comment to oapv_version function for Android versioning

Signed-off-by: Dawid Kozinski <d.kozinski@samsung.com>

dkozinski committed:
Update versioning scheme in CMakeLists.txt and pkgconfig file

- Changed CPACK_PACKAGE_VERSION to use VERSION_APISET, VERSION_MAJOR, VERSION_MINOR, and VERSION_PATCH for better version management.
- Updated oapv.pc.in to reflect the new versioning format.

Signed-off-by: Dawid Kozinski <d.kozinski@samsung.com>

dkozinski committed:
Refactor versioning messages and declarations

Signed-off-by: Dawid Kozinski <d.kozinski@samsung.com>

kpchoi committed:
refactoring code to avoid a symbol conflict. moved version string conversion code to application

Signed-off-by: kp5.choi@samsung.com <kp5.choi@samsung.com>

kpchoi commented on inc/oapv.h:
```diff
@@ -41,29 +41,20 @@ extern "C" {
-        buffer; \
-    })
+/* the version string follows the rule of API_SET.MAJOR.MINOR.PATCH */
+#define OAPV_VER_SET(apiset, major, minor, patch) \
```
I think that the macro sets a version number with four numbers.
It is reason why I used 'SET'.

kpchoi commented on app/oapv_app_util.h:
```diff
@@ -789,4 +789,16 @@ static unsigned char char_to_hex(char a)
 }
 
+/* get version string */
+char * ver_num_2_ver_str(unsigned int ver_num)
```
The funciton name will be changed to ```ver_num_to_str()```.

kpchoi committed:
the candidate of consensus

Signed-off-by: kp5.choi@samsung.com <kp5.choi@samsung.com>

dkozinski committed:
refactor CMakeLists.txt: Extract version information from oapv.h instead of version.txt

- Removed the function for retrieving version from git tags and version.txt.
- Added checks to ensure the existence of oapv.h and extract version information using regex.
- Updated project versioning to use values defined in oapv.h.

Signed-off-by: Dawid Kozinski <d.kozinski@samsung.com>

jamrial commented on src/oapv.c:
```diff
@@ -2153,5 +2153,11 @@ int oapvd_info(void *au, int au_size, oapv_au_info_t *aui)
 
-static char *oapv_ver = "0.1.13.1";
-char * oapv_version() { return oapv_ver; }
+char *oapv_version(void)
```
Can't this be made a `const char*` instead? With CMake it's trivial to generate the "0.1.13.1" string at configure time, but i assume the issue is with Android?
As is, this looks potentially racy. If it can't be const, then it should allocate a buffer and return it, and the user should then free it.

Also, an int version is still extremely useful for runtime checks, so please add it as a separate function, like `unsigned int oapv_version_int()`

jamrial replied:
To generate the int to be returned? Yes

```
unsigned int oapv_version_int()
{
    return OAPV_VER_SET(OAPV_VER_APISET, OAPV_VER_MAJOR, OAPV_VER_MINOR, OAPV_VER_PATCH);
}
```

Without a function like this, API users that want the version in integer form would need to parse the output of oapv_version()

jamrial replied:
The function may return not just the version number but also an error code, which is a negative value, despite the return type being unsigned, and as such can't be checked at all. For all the API user knows, it just returned a valid version number they can parse with OAPV_VER_GET_*.

kpchoi replied:
> The function may return not just the version number but also an error code, which is a negative value, despite the return type being unsigned, and as such can't be checked at all. For all the API user knows, it just returned a valid version number they can parse with OAPV_VER_GET_*.

Again, oapv_version() has changed to support your comments.
Check the PR https://github.com/AcademySoftwareFoundation/openapv/pull/105 .

dkozinski committed:
Further refactoring related to version handling

- Removed the inclusion of oapv_version.h in oapv.h.
- Changed the return type of oapv_version function from char* to const char*.
- Deleted the oapv_version.h.in file as it is no longer needed.
- Updated CMakeLists.txt to remove references to the version header file.

Signed-off-by: Dawid Kozinski <d.kozinski@samsung.com>

kpchoi reviewed (approved):
LGTM

cpncf commented:
It has been long discussion here.
I'd like to strongly recommend to merge this PR of the current version as soon as possible.

dkozinski committed:
Added oapv_version_int function that returns the version as an unsigned int

Signed-off-by: Dawid Kozinski <d.kozinski@samsung.com>

dariusz-f reviewed (approved):
LGTM

dkozinski committed:
Removed oapv_version_int function that returns the version as an unsigned int

-The function was redundant because there is a macro OAPV_VER_SET that does this

Signed-off-by: Dawid Kozinski <d.kozinski@samsung.com>
